### PR TITLE
[CHERI-RISC-V][compiler-rt] CHERI-aware save.S/restore.S

### PR DESCRIPTION
**Pull request by heshamelmatary**
This makes compiler-rt builds for purecap code again. Not tested in run-time.

### CODE REVIEW

File: compiler-rt/lib/builtins/riscv/restore.S [hunk 1]
```diff
@@ -79,20 +100,34 @@ __riscv_restore_3:
 __riscv_restore_2:
 __riscv_restore_1:
 __riscv_restore_0:
+#ifdef __CHERI_PURE_CAPABILITY__
+  clc      cs2, 0(csp)
```
Comment (arichardson) [inline compiler-rt/lib/builtins/riscv/restore.S:104]: Those offsets are wrong?

Comment (heshamelmatary) [reply compiler-rt/lib/builtins/riscv/restore.S:104]: 0 is correct, but the following 3 offsets are wrong, I fixed them.

**Comment by jrtc27** [comment]
The upstream code very carefully ensures the minimum stack usage, only grouping together spills and reloads that just fill up wasted stack bytes due to alignment (i.e. the grouping factor is `16 / sizeof(register)`). That means purecap RV32 should have the same grouping as vanilla RV64, and purecap RV64 should have no grouping whatsoever.

**Comment by heshamelmatary** [comment]
> The upstream code very carefully ensures the minimum stack usage, only grouping together spills and reloads that just fill up wasted stack bytes due to alignment (i.e. the grouping factor is `16 / sizeof(register)`). That means purecap RV32 should have the same grouping as vanilla RV64, and purecap RV64 should have no grouping whatsoever.

Makes more sense. I will refactor the PR to do so.

**Comment by jrtc27** [comment]
#571

**Pull request closed without merging**